[LT] Add a trie data structure for caching IR nodes

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76738
* #76711
* __->__ #76542

Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

desertfire committed:
[LT] Add a trie data structure for caching IR nodes

Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

wconstab commented on test/cpp/lazy/test_trie_cache.cpp:
```diff
@@ -0,0 +1,88 @@
+  NodePtr c = MakeNode<TrieCacheNode>(2);
+  TrieCache::Get()->ResetCurrent(); // MarkStep
+
+  EXPECT_EQ(ReuseOrMakeNode<TrieCacheNode>(OpKind(), 0).get(), a.get());
```
i was surprised you can pass OpKind() here if it is needed for casting, so i looked at NodeCast.  It makes sense now.. no need to change it for this PR.

My idea for cleaning up later is to make opkind a static member of the derived class rather than a field of node base.  Then, could we change NodeCast<T>(node) to access T.kind directly?  (not sure if that is valid in template code)

desertfire replied:
Agree that we need some way to carry the opcode at the class level. But as you pointed out the template code makes it difficult. I will update this once I figure out a good way to do it.

desertfire replied:
Currently the best I can do is to add a static member for each derived class, and make the base Node op() method virtual so that it can return the correct op for each derived class. That will be a fairly big change and definitely worth to be separated in another PR.

wconstab reviewed (approved):
lgtm thanks @desertfire

JackCaoG reviewed (approved):
LGTM, thanks!

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge this

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add a trie data structure for caching IR nodes"


Summary: TrieCache provides a way to look up an IR node before we
actually create it. If the lookup hits in TrieCache, we reuse the
existing node and move the current pointer in TrieCache to point to that
node; if the lookup misses, we create a new node and insert it into TrieCache.

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge this
